man/cups-oauth.1
.\"
.\" cups-oauth man page for CUPS.
.\"
.\" Copyright © 2025 by OpenPrinting.
.\"
.\" Licensed under Apache License v2.0.  See the file "LICENSE" for more
.\" information.
.\"
.TH cups-oauth 1 "CUPS" "2025-05-05" "OpenPrinting"
.SH NAME
cups-oauth \- interact with an oauth/openid authorization server
.SH SYNOPSIS
.B cups-oauth
.B \-\-help
.br
.B cups-oauth
.B \-\-version
.br
.B cups-oauth
[
.B \-a
.I OAUTH-URI
] [
.B \-s
.I SCOPE(S)
]
.B authorize
.I [RESOURCE]
.br
.B cups-oauth
[
.B \-a
.I OAUTH-URI
]
.B clear
.I [RESOURCE]
.br
.B cups-oauth
[
.B \-a
.I OAUTH-URI
]
.B get-access-token
.I [RESOURCE]
.br
.B cups-oauth
[
.B \-a
.I OAUTH-URI
]
.B get-client-id
.br
.B cups-oauth
[
.B \-a
.I OAUTH-URI
]
.B get-metadata
.I [NAME]
.br
.B cups-oauth
[
.B \-a
.I OAUTH-URI
]
.B get-user-id
.I [RESOURCE]
.I [NAME]
.br
.B cups-oauth
[
.B \-a
.I OAUTH-URI
]
.B set-access-token
.I [RESOURCE]
.I TOKEN
.br
.B cups-oauth
[
.B \-a
.I OAUTH-URI
]
.B set-client-data
.I CLIENT-ID
.I CLIENT-SECRET
.SH DESCRIPTION
The
.B cups-oauth
utility interacts with an OAuth/OpenID authorization server.
Authorizations are often linked to a resource (a printer URI, web page URL, etc.)
.SH OPTIONS
The following options are recognized by
.B cups-oauth:
.TP 5
.B \-\-help
Show program usage.
.TP 5
.B \-\-version
Show the CUPS version.
.TP 5
\fB\-a \fIOAUTH-URI\fR
Specifies the OAuth/OpenID authorization server URL.
.TP 5
\fB\-s \fISCOPE(S)\fR
Specifies a space-delimited list of scope names to use when authorizing access.
The default is to request authorization for all supported OpenID scopes.
.SH SUB-COMMANDS
.SS authorize
Starts an authorization workflow with the default web browser.
If a resource URI is specified, the authorization is specific to that resource.
The access token is send to the standard output on success.
.SS clear
Clears any authorization for the specified resource or for all resources if no resource URI is supplied.
.SS get-access-token
Output the current, unexpired access token, if any, to the standard output.
.SS get-client-id
Output the client ID string, if any, to the standard output.
.SS get-metadata
Get the OAuth/OpenID authorization server metadata and send it to the standard output.
If a name is specified, the output is just the value for the specified metadata.
.SS get-user-id
Get the OpenID user ID information and send it to the standard output.
If a name is specified, the output is just the named claim from the user ID.
.SS set-access-token
Set the access token (which is sometimes also called an API key) for the specified resource or for all resources.
.SS set-client-data
Set the client ID string and secret for an OAuth/OpenID authorization server.
.SH ENVIRONMENT VARIABLES
The
.B CUPS_OAUTH_URI
environment variable sets the default OAuth/OpenID authorization server URL.
.PP
The
.B CUPS_OAUTH_SCOPES
environment variable sets the default OAuth/OpenID scopes as a space-delimited list.
.SH NOTES
CUPS uses a redirect URI of "http://127.0.0.1/" for all authorization on the local system.
.SH EXAMPLES
Register a client ID and secret for the OAuth server at "https://oauth.example.com/":
.nf
     cups-oauth -a https://oauth.example.com/ set-client-data CLIENT-ID CLIENT-SECRET
.fi
Save an access token (sometimes called an application or API key) for the OAuth server at "https://oauth.example.com/":
.nf
     cups-oauth -a https://oauth.example.com/ set-access-token TOKEN
.fi
Authorize against the OAuth server at "https://oauth.example.com/" using your web browser:
.nf
     cups-oauth -a https://oauth.example.com/ authorize
.fi
.SH SEE ALSO
.BR cups (1)
.SH COPYRIGHT
Copyright \[co] 2025 by OpenPrinting.


CHANGES.md
CHANGES - OpenPrinting CUPS
===========================


Changes in CUPS v2.4.13 (YYYY-MM-DD)
------------------------------------

- Updated the scheduler to send the "printer-added" or "printer-modified" events
  whenever an IPP Everywhere PPD is installed (Issue #1244)
- Updated the scheduler to send the "printer-modified" event whenever the system
  default printer is changed (Issue #1246)
- Fixed a memory leak in `httpClose` (Issue #1223)
- Fixed missing commas in `ippCreateRequestedArray` (Issue #1234)
- Fixed subscription issues in the scheduler and D-Bus notifier (Issue #1235)
- Fixed support for IPP/PPD options with periods or underscores (Issue #1249)


Changes in CUPS v2.4.12 (2025-04-08)
------------------------------------

- GnuTLS follows system crypto policies now (Issue #1105)
- Added `NoSystem` SSLOptions value (Issue #1130)
- Now we raise alert for certificate issues (Issue #1194)
- Added Kyocera USB quirk (Issue #1198)
- The scheduler now logs a job's debugging history if the backend fails
  (Issue #1205)
- Fixed a potential timing issue with `cupsEnumDests` (Issue #1084)
- Fixed a potential "lost PPD" condition in the scheduler (Issue #1109)
- Fixed a compressed file error handling bug (Issue #1070)
- Fixed a bug in the make-and-model whitespace trimming code (Issue #1096)
- Fixed a removal of IPP Everywhere permanent queue if installation failed
  (Issue #1102)
- Fixed `ServerToken None` in scheduler (Issue #1111)
- Fixed invalid IPP keyword values created from PPD option names (Issue #1118)
- Fixed handling of "media" and "PageSize" in the same print request
  (Issue #1125)
- Fixed client raster printing from macOS (Issue #1143)
- Fixed the default User-Agent string.
- Fixed a recursion issue in `ippReadIO`.
- Fixed handling incorrect radix in `scan_ps()` (Issue #1188)
- Fixed validation of dateTime values with time zones more than UTC+11
  (Issue #1201)
- Fixed attributes returned by the Create-Xxx-Subscriptions requests
  (Issue #1204)
- Fixed `ippDateToTime` when using a non GMT/UTC timezone (Issue #1208)
- Fixed `job-completed` event notifications for jobs that are cancelled before
  started (Issue #1209)
- Fixed DNS-SD discovery with `ippfind` (Issue #1211)


Changes in CUPS v2.4.11 (2024-09-30)
------------------------------------

- Updated the maximum file descriptor limit for `cupsd` to 64k-1 (Issue #989)
- Fixed `lpoptions -d` with a discovered but not added printer (Issue #833)
- Fixed incorrect error message for HTTP/IPP errors (Issue #893)
- Fixed JobPrivateAccess and SubscriptionPrivateAccess support for "all"
  (Issue #990)
- Fixed issues with cupsGetDestMediaByXxx (Issue #993)
- Fixed adding and modifying of printers via the web interface (Issue #998)
- Fixed HTTP PeerCred authentication for domain users (Issue #1001)
- Fixed checkbox support (Issue #1008)
- Fixed printer state notifications (Issue #1013)
- Fixed IPP Everywhere printer setup (Issue #1033)


Changes in CUPS v2.4.10 (2024-06-18)
------------------------------------

- Fixed error handling when reading a mixed `1setOf` attribute.
- Fixed scheduler start if there is only domain socket to listen on (Issue #985)


Changes in CUPS v2.4.9 (2024-06-11)
-----------------------------------

- Fixed domain socket handling (CVE-2024-35235)
- Fixed creating of `cupsUrfSupported` PPD keyword (Issue #952)
- Fixed searching for destinations in web ui (Issue #954)
- Fixed TLS negotiation using OpenSSL with servers that require the TLS SNI
  extension.
- Really raised `cups_enum_dests()` timeout for listing available IPP printers
  (Issue #751)...
- Fixed `Host` header regression (Issue #967)
- Fixed DNS-SD lookups of local services with Avahi (Issue #970)
- Fixed listing jobs in destinations in web ui. (Apple issue #6204)
- Fixed showing search query in web ui help page. (Issue #977)


Changes in CUPS v2.4.8 (2024-04-26)
-----------------------------------

- Added warning if the device has to be asked for 'all,media-col-database'
  separately (Issue #829)
- Added new value for 'lpstat' option '-W' - successfull - for getting
  successfully printed jobs (Issue #830)
- Added support for PAM modules password-auth and system-auth (Issue #892)
- Updated IPP Everywhere printer creation error reporting (Issue #347)
- Updated and documented the MIME typing buffering limit (Issue #925)
- Now report an error for temporary printer defaults with lpadmin (Issue #237)
- Fixed mapping of PPD InputSlot, MediaType, and OutputBin values (Issue #238)
- Fixed "document-unprintable-error" handling (Issue #391)
- Fixed the web interface not showing an error for a non-existent printer
  (Issue #423)
- Fixed printing of jobs with job name longer than 255 chars on older printers
  (Issue #644)
- Really backported fix for Issue #742
- Fixed `cupsCopyDestInfo` device connection detection (Issue #586)
- Fixed "Upgrade" header handling when there is no TLS support (Issue #775)
- Fixed memory leak when unloading a job (Issue #813)
- Fixed memory leak when creating color profiles (Issue #815)
- Fixed a punch finishing bug in the IPP Everywhere support (Issue #821)
- Fixed crash in `scan_ps()` if incoming argument is NULL (Issue #831)
- Fixed setting job state reasons for successful jobs (Issue #832)
- Fixed infinite loop in IPP backend if hostname is IP address with Kerberos
  (Issue #838)
- Added additional check on socket if `revents` from `poll()` returns POLLHUP
  together with POLLIN or POLLOUT in `httpAddrConnect2()` (Issue #839)
- Fixed crash in `ppdEmitString()` if `size` is NULL (Issue #850)
- Fixed reporting `media-source-supported` when sharing printer which has
  numbers as strings instead of keywords as `InputSlot` values (Issue #859)
- Fixed IPP backend to support the "print-scaling" option with IPP printers
  (Issue #862)
- Fixed potential race condition for the creation of temporary queues
  (Issue #871)
- Fixed `httpGets` timeout handling (Issue #879)

tools/cups-oauth.c
//
// OAuth utility for CUPS.
//
// Copyright © 2024-2025 by OpenPrinting.
//
// Licensed under Apache License v2.0.  See the file "LICENSE" for more
// information.
//
// Usage: cups-oauth [OPTIONS] [SUB-COMMAND [ARGUMENT(S)]]
//
// Commands:
//
//   authorize [RESOURCE]
//   clear [RESOURCE]
//   get-access-token [RESOURCE]
//   get-client-id
//   get-metadata [NAME]
//   get-refresh-token [RESOURCE]
//   get-user-id [RESOURCE] [NAME]
//   set-access-token [RESOURCE] TOKEN
//   set-client-data CLIENT-ID CLIENT-SECRET
//
// Options:
//
//   --help
//   --version
//   -a OAUTH-URI
//   -s SCOPE(S)
//

#include <cups/cups-private.h>
#include <cups/oauth.h>


//
// Macro for localized text...
//

#  define _(x) x


//
// Local functions...
//

static int	do_authorize(const char *oauth_uri, const char *scopes, const char *resource_uri);
static int	do_clear(const char *oauth_uri, const char *resource_uri);
static int	do_get_access_token(const char *oauth_uri, const char *resource_uri);
static int	do_get_client_id(const char *oauth_uri);
static int	do_get_metadata(const char *oauth_uri, const char *name);
static int	do_get_user_id(const char *oauth_uri, const char *resource_uri, const char *name);
static int	do_set_access_token(const char *oauth_uri, const char *resource_uri, const char *token);
static int	do_set_client_data(const char *oauth_uri, const char *client_id, const char *client_secret);
static int	usage(FILE *out);


//
// 'main()' - Main entry.
//

int					// O - Exit status
main(int  argc,				// I - Number of command-line arguments
     char *argv[])			// I - Command-line arguments
{
  int		i;			// Looping var
  const char	*opt,			// Current option
		*oauth_uri = getenv("CUPS_OAUTH_URI"),
					// OAuth authorization server URI
		*scopes = getenv("CUPS_OAUTH_SCOPES");
					// Scopes


  // Parse the command-line...
  for (i = 1; i < argc; i ++)
  {
    if (!strcmp(argv[i], "--help"))
    {
      return (usage(stdout));
    }
    else if (!strcmp(argv[i], "--version"))
    {
      puts(CUPS_SVERSION);
      exit(0);
    }
    else if (!strncmp(argv[i], "--", 2))
    {
      _cupsLangPrintf(stderr, _("%s: Unknown option '%s'."), "cups-oauth", argv[i]);
      return (usage(stderr));
    }
    else if (argv[i][0] == '-' && argv[i][1] != '-')
    {
      for (opt = argv[i] + 1; *opt; opt ++)
      {
        switch (*opt)
        {
          case 'a' : // -a AUTH-URI
              i ++;
              if (i >= argc)
              {
                _cupsLangPuts(stderr, _("cups-oauth: Missing Authorization Server URI after '-a'."));
                return (usage(stderr));
              }

              oauth_uri = argv[i];
              break;

          case 's' : // -s SCOPE(S)
              i ++;
              if (i >= argc)
              {
                _cupsLangPuts(stderr, _("cups-oauth: Missing scope(s) after '-s'."));
                return (usage(stderr));
              }

              scopes = argv[i];
              break;

          default :
              _cupsLangPrintf(stderr, _("%s: Unknown option '-%c'."), "cups-oauth", *opt);
              return (usage(stderr));
        }
      }
    }
    else if (!oauth_uri)
    {
      _cupsLangPuts(stderr, _("cups-oauth: No authorization server specified."));
      return (usage(stderr));
    }
    else if (!strcmp(argv[i], "authorize"))
    {
      // authorize [RESOURCE]
      i ++;
      return (do_authorize(oauth_uri, scopes, argv[i]));
    }
    else if (!strcmp(argv[i], "clear"))
    {
      // clear [RESOURCE]
      i ++;
      return (do_clear(oauth_uri, argv[i]));
    }
    else if (!strcmp(argv[i], "get-access-token"))
    {
      // get-access-token [RESOURCE]
      i ++;
      return (do_get_access_token(oauth_uri, argv[i]));
    }
    else if (!strcmp(argv[i], "get-client-id"))
    {
      // get-client-id
      i ++;
      return (do_get_client_id(oauth_uri));
    }
    else if (!strcmp(argv[i], "get-metadata"))
    {
      // get-metadata [NAME]
      i ++;
      return (do_get_metadata(oauth_uri, argv[i]));
    }
    else if (!strcmp(argv[i], "get-user-id"))
    {
      // get-user-id [RESOURCE] [NAME]
      i ++;
      if (i < argc)
      {
        if (!strncmp(argv[i], "ipp://", 6) || !strncmp(argv[i], "ipps://", 7) || !strncmp(argv[i], "http://", 7) || !strncmp(argv[i], "https://", 8))
	  return (do_get_user_id(oauth_uri, argv[i], argv[i + 1]));
        else
	  return (do_get_user_id(oauth_uri, /*resource_uri*/NULL, argv[i]));
      }
      else
      {
        return (do_get_user_id(oauth_uri, /*resource_uri*/NULL, /*name*/NULL));
      }
    }
    else if (!strcmp(argv[i], "set-access-token"))
    {
      // set-access-token [RESOURCE] TOKEN
      i ++;
      if (i >= argc)
      {
	_cupsLangPuts(stderr, _("cups-oauth: Missing resource URI and/or access token."));
	return (usage(stderr));
      }

      return (do_set_access_token(oauth_uri, argv[i], argv[i + 1]));
    }
    else if (!strcmp(argv[i], "set-client-data"))
    {
      // set-client-data CLIENT-ID CLIENT-DATA
      i ++;
      if ((i + 1) >= argc)
      {
	_cupsLangPuts(stderr, _("cups-oauth: Missing client_id and/or client_secret."));
	return (usage(stderr));
      }

      return (do_set_client_data(oauth_uri, argv[i], argv[i + 1]));
    }
    else
    {
      _cupsLangPrintf(stderr, _("cups-oauth: Unknown command '%s'."), argv[i]);
      return (usage(stderr));
    }
  }

  // If we get this far, show usage...
  return (usage(argc == 1 ? stdout : stderr));
}


//
// 'do_authorize()' - Authorize access.
//

static int				// O - Exit status
do_authorize(const char *oauth_uri,	// I - Authorization Server URI
             const char *scopes,	// I - Scope(s)
             const char *resource_uri)	// I - Resource URI
{
  int		status = 1;		// Exit status
  cups_json_t	*metadata;		// Server metadata
  char		*auth_code = NULL,	// Authorization code
		*access_token = NULL;	// Access token
  time_t	access_expires;		// Expiration date


  // Get the server metadata...
  if ((metadata = cupsOAuthGetMetadata(oauth_uri)) == NULL)
  {
    _cupsLangPrintf(stderr, _("cups-oauth: Unable to get metadata for '%s': %s"), oauth_uri, cupsGetErrorString());
    return (1);
  }

  // Authorize...
  if ((auth_code = cupsOAuthGetAuthorizationCode(oauth_uri, metadata, resource_uri, scopes, /*redirect_uri*/NULL)) == NULL)
  {
    _cupsLangPrintf(stderr, _("cups-oauth: Unable to get authorization from '%s': %s"), oauth_uri, cupsGetErrorString());
    goto done;
  }

  // Get the access token...
  if ((access_token = cupsOAuthGetTokens(oauth_uri, metadata, resource_uri, auth_code, CUPS_OGRANT_AUTHORIZATION_CODE, CUPS_OAUTH_REDIRECT_URI, &access_expires)) == NULL)
  {
    _cupsLangPrintf(stderr, _("cups-oauth: Unable to get access token from '%s': %s"), oauth_uri, cupsGetErrorString());
    goto done;
  }

  // Show access token
  puts(access_token);

  status = 0;

  // Clean up and return...
  done:

  cupsJSONDelete(metadata);
  free(auth_code);
  free(access_token);

  return (status);
}


//
// 'do_clear()' - Clear authorization information.
//

static int				// O - Exit status
do_clear(const char *oauth_uri,		// I - Authorization Server URI
         const char *resource_uri)	// I - Resource URI
{
  cupsOAuthClearTokens(oauth_uri, resource_uri);

  return (0);
}


//
// 'do_get_access_token()' - Get an access token.
//

static int				// O - Exit status
do_get_access_token(
    const char *oauth_uri,		// I - Authorization Server URI
    const char *resource_uri)		// I - Resource URI
{
  char		*access_token;		// Access token
  time_t	access_expires;		// Expiration date


  if ((access_token = cupsOAuthCopyAccessToken(oauth_uri, resource_uri, &access_expires)) != NULL)
  {
    puts(access_token);
    free(access_token);
    return (0);
  }

  return (1);
}


//
// 'do_get_client_id()' - Get the client ID value.
//

static int				// O - Exit status
do_get_client_id(
    const char *oauth_uri)		// I - Authorization Server URI
{
  char	*client_id;			// Client ID


  if ((client_id = cupsOAuthCopyClientId(oauth_uri, CUPS_OAUTH_REDIRECT_URI)) != NULL)
  {
    puts(client_id);
    free(client_id);
    return (0);
  }

  return (1);
}


//
// 'do_get_metadata()' - Get authorization server metadata.
//

static int				// O - Exit status
do_get_metadata(const char *oauth_uri,	// I - Authorization Server URI
                const char *name)	// I - Field name
{
  cups_json_t	*metadata;		// Metadata
  char		*json;			// JSON string


  // Get the metadata...
  if ((metadata = cupsOAuthGetMetadata(oauth_uri)) == NULL)
  {
    _cupsLangPrintf(stderr, _("cups-oauth: Unable to get metadata for '%s': %s"), oauth_uri, cupsGetErrorString());
    return (1);
  }

  // Show metadata...
  if (name)
  {
    cups_json_t	*value = cupsJSONFind(metadata, name);
					// Metadata value

    if (value)
    {
      switch (cupsJSONGetType(value))
      {
        case CUPS_JTYPE_NULL :
            puts("null");
            break;

        case CUPS_JTYPE_FALSE :
            puts("false");
            break;

        case CUPS_JTYPE_TRUE :
            puts("true");
            break;

        case CUPS_JTYPE_NUMBER :
            printf("%g\n", cupsJSONGetNumber(value));
            break;

        case CUPS_JTYPE_STRING :
            puts(cupsJSONGetString(value));
            break;

        default :
            if ((json = cupsJSONExportString(value)) != NULL)
            {
              puts(json);
              free(json);
            }
            break;
      }

      return (0);
    }
    else
    {
      return (1);
    }
  }
  else if ((json = cupsJSONExportString(metadata)) != NULL)
  {
    puts(json);
    free(json);
  }

  return (0);
}


//
// 'do_get_user_id()' - Get user identification.
//

static int				// O - Exit status
do_get_user_id(
    const char *oauth_uri,		// I - Authorization Server URI
    const char *resource_uri,		// I - Resource URI
    const char *name)			// I - Claim name
{
  cups_jwt_t	*user_id;		// User ID information
  cups_json_t	*claims;		// Claims
  char		*json,			// JSON string
		date[256];		// Date
  int		ret = 0;		// Exit status


  // Get the user_id...
  if ((user_id = cupsOAuthCopyUserId(oauth_uri, resource_uri)) == NULL)
  {
    _cupsLangPrintf(stderr, _("cups-oauth: Unable to get user ID for '%s': %s"), oauth_uri, cupsGetErrorString());
    return (1);
  }

  claims = cupsJWTGetClaims(user_id);

  // Show user information...
  if (name)
  {
    cups_json_t	*value = cupsJSONFind(claims, name);
					// Claim value

    if (value)
    {
      switch (cupsJSONGetType(value))
      {
        case CUPS_JTYPE_NULL :
            puts("null");
            break;

        case CUPS_JTYPE_FALSE :
            puts("false");
            break;

        case CUPS_JTYPE_TRUE :
            puts("true");
            break;

        case CUPS_JTYPE_NUMBER :
            if (!strcmp(name, "exp") || !strcmp(name, "iat") || !strcmp(name, "nbf"))
              puts(httpGetDateString2((time_t)cupsJSONGetNumber(value), date, sizeof(date)));
            else
	      printf("%g\n", cupsJSONGetNumber(value));
            break;

        case CUPS_JTYPE_STRING :
            puts(cupsJSONGetString(value));
            break;

        default :
            if ((json = cupsJSONExportString(value)) != NULL)
            {
              puts(json);
              free(json);
            }
            break;
      }
    }
    else
    {
      ret = 1;
    }
  }
  else if ((json = cupsJSONExportString(claims)) != NULL)
  {
    puts(json);
    free(json);
  }

  cupsJWTDelete(user_id);

  return (ret);
}


//
// 'do_set_access_token()' - Set the access token.
//

static int				// O - Exit status
do_set_access_token(
    const char *oauth_uri,		// I - Authorization Server URI
    const char *resource_uri,		// I - Resource URI
    const char *token)			// I - Access token
{
  cupsOAuthSaveTokens(oauth_uri, resource_uri, token, /*access_expires*/time(NULL) + 365 * 86400, /*user_id*/NULL, /*refresh_token*/NULL);

  return (0);
}


//
// 'do_set_client_data()' - Save client_id and client_secret values.
//

static int				// O - Exit status
do_set_client_data(
    const char *oauth_uri,		// I - Authorization Server URI
    const char *client_id,		// I - Client ID
    const char *client_secret)		// I - Client secret
{
  cupsOAuthSaveClientData(oauth_uri, CUPS_OAUTH_REDIRECT_URI, client_id, client_secret);

  return (0);
}


//
// 'usage()' - Show usage.
//

static int				// O - Exit status
usage(FILE *out)			// I - Output file
{
  _cupsLangPuts(out, _("Usage: cups-oauth [OPTIONS] [SUB-COMMAND [ARGUMENT(S)]]"));
  _cupsLangPuts(out, "");
  _cupsLangPuts(out, _("Sub-Commands:"));
  _cupsLangPuts(out, "");
  _cupsLangPuts(out, _("authorize [RESOURCE]           Authorize access to a resource"));
  _cupsLangPuts(out, _("clear [RESOURCE]               Clear the authorization for a resource"));
  _cupsLangPuts(out, _("get-access-token [RESOURCE]    Get the current access token"));
  _cupsLangPuts(out, _("get-client-id                  Get the client ID for the authorization server"));
  _cupsLangPuts(out, _("get-metadata [NAME]            Get metadata from the authorization server"));
  _cupsLangPuts(out, _("get-user-id [RESOURCE] [NAME]  Get the authorized user ID"));
  _cupsLangPuts(out, _("set-access-token [RESOURCE] TOKEN\n"
                      "                               Set the current access token"));
  _cupsLangPuts(out, _("set-client-data CLIENT-ID CLIENT-SECRET\n"
                      "                               Set the client ID and secret for the authorization server."));
  _cupsLangPuts(out, "");
  _cupsLangPuts(out, _("Options:"));
  _cupsLangPuts(out, "");
  _cupsLangPuts(out, _("--help                         Show this help"));
  _cupsLangPuts(out, _("--version                      Show the program version"));
  _cupsLangPuts(out, _("-a OAUTH-URI                   Specify the OAuth authorization server URL"));
  _cupsLangPuts(out, _("-s SCOPE(S)                    Specify the scope(s) to authorize"));

  return (out == stdout ? 0 : 1);
}
